Merging robot relative drive into dev

BrandonBeaver117 opened the pull request:
Adding robot relative drive suppor

BrandonBeaver117 committed:
First attempt at robot-oriented drive

Made new command that's basically  ignoring gyro, hopefully making it robot-relative

BrandonBeaver117 committed:
Better robot oriented driving

Modified driving commands. If right bumper pressed, ignore gyro and can drive robot relative. If right bumper released, then it goes back to field relative.

BrandonBeaver117 committed:
Rotation snapping robot relative

Robot now can snap to robot-relative angle. Adds current angle to target angle if robot-relative angle needed. Hold down right bumper to enable robot relative mode

BrandonBeaver117 committed:
Robot relative angle snapping

Hold right bumper, angle snapping is now robot relative. Cannot perform simultaneous snap commands, as the command has to finish first before initializing a new start angle

a-viv-a commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -57,7 +57,7 @@ public static final class FrontRight { // Module 1
 
         public static final double STEER_OFFSET =
             IS_COMP_BOT
-                ? -Math.toRadians(329.4937) // comp bot offset
```
Why are you setting offsets in a feature?

BrandonBeaver117 replied:
I wanted to test its driving, but modules needed calibration first. In hindsight, probably should have made a fix-branch first (like you said)

a-viv-a commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -68,7 +68,7 @@ public static final class FrontLeft { // Module 2
         public static final double STEER_OFFSET =
             IS_COMP_BOT
-                ? -Math.toRadians(30.498) // comp bot offset
+                ? -Math.toRadians(331.41632080078125) // comp bot offset
```
Why are you setting offsets in a feature? This is fine, but it's better as a fix branch -- it can be merged independently of your feature, so it should be independent.

a-viv-a commented on src/main/java/frc/robot/commands/RotateVectorDriveCommand.java:
```diff
@@ -24,8 +25,12 @@ public class RotateVectorDriveCommand extends CommandBase {
   private final DoubleSupplier rotationYSupplier;
+  private final BooleanSupplier isRobotRelativeSupplier;
+
+  private double initialAngle;
```
a doccomment about what this member stores would be nice.

a-viv-a commented on src/main/java/frc/robot/commands/RotateVectorDriveCommand.java:
```diff
@@ -57,10 +65,17 @@ public void execute() {
     double rotY = rotationYSupplier.getAsDouble();
+    boolean isRobotRelative = isRobotRelativeSupplier.getAsBoolean();
 
     // if stick magnitude is greater then rotate angle mag
```
this comment stayed in the same place, but the code it was referring to moved -- please dont do that

a-viv-a commented on src/main/java/frc/robot/commands/RotateVectorDriveCommand.java:
```diff
@@ -57,10 +65,17 @@ public void execute() {
+    if (Util.vectorMagnitude(rotX, rotY) > Drive.ROTATE_VECTOR_MAGNITUDE && !isRobotRelative) {
+      angle = targetAngle;
+    } else if (Util.vectorMagnitude(rotX, rotY) > Drive.ROTATE_VECTOR_MAGNITUDE
+        && isRobotRelative) {
```
ths boolean logic is reduntent and fairly ugly -- it could be expressed with a single if and a ternary.
`` if (Util.vectorMagnitude(rotX, rotY) > Drive.ROTATE_VECTOR_MAGNITUDE) ``
and then ``angle = isRobotRelative ? /** ect ect */``

BrandonBeaver117 replied:
yeah that looks better

a-viv-a commented on src/main/java/frc/robot/commands/RotateVelocityDriveCommand.java:
```diff
@@ -20,18 +22,24 @@ public class RotateVelocityDriveCommand extends CommandBase {
+    this.isRobotRelativeSupplier = isRobotRelativeSupplier;
 
+    Shuffleboard.getTab("Drivebase")
+        .addBoolean("Is robot relative drive?", isRobotRelativeSupplier);
```
if this is just for testing, can we comment it out before we merge it? each shuffleboard value adds network traffic and cpu load.

a-viv-a reviewed (changes requested):
needs a bit of work, but overall seems good -- nice work!

BrandonBeaver117 committed:
Fix pull review feedback

Moving comments, commenting out shuffleboard value, improving rotation logic cleanliness, adding comment about intitial angle

BrandonBeaver117 committed:
rolling back constants change

BrandonBeaver117 committed:
rolling back navx for autos

a-viv-a commented:
@BrandonTheBusyBeaver117 this branch is out of date -- you need to merge `dev` -> `feat/robot-rel` for the module offset changes. Github provides a helpful button for it below.

BrandonBeaver117 committed:
Merge branch 'dev' into feat/robot-rel

a-viv-a reviewed (approved):
LGTM
feel free to merge if you want Brandon, or get a review from Lexie / Audrey

BrandonBeaver117 committed:
Merge branch 'dev' into feat/robot-rel
